### PR TITLE
Add new `use-ember-data-rfc-395-imports` rule

### PR DESCRIPTION
**Pull request by dcyriller**
## Description

This PR introduces a new rule to ease ember-data migration to new packages: `use-ember-data-rfc-395-imports`.

Examples of **incorrect** code for this rule:
```js
import Model from "ember-data/model";
import attr from "ember-data/attr";

import DS from "ember-data";
const { Model } = "ember-data";
```

Examples of **correct** code for this rule:

```js
import Model, { attr } from "@ember-data/model";

import Model from "@ember-data/model";
```

## Note

Split into 5 commits that may be reviewed one after the other.

It leverages what have been done in the past with Ember globals and ember-cli-shims.

~~A few discussion are still in progress around https://github.com/ember-data/ember-data-rfc395-data/ (the data set). Hence the WIP!~~

## Links 

[ember-data RFC](https://github.com/ember-data/ember-data-rfc395-data)
[a related codemod](https://github.com/ember-codemods/ember-data-codemod)

_EDIT: I've updated this description to reflect the changes made following https://github.com/ember-data/ember-data-rfc395-data/pull/7_

### CODE REVIEW

**Comment by dcyriller** [comment]
cc @runspired @rwjblue

File: docs/rules/new-ember-data-packages.md [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+## How to fix
+
+This rule implements a fix function (or to be more precise: it leverages the `no-old-shims` fix function). If, as a user, you can get rid of all the `import DS from "ember-data";` imports, you'll be able to use this rule fixer to complete the upgrade. 
+
+Note that a codemod is also available to complete the upgrade.
+
+## When Not To Use It
+
+You don't want to use this rule if you're using `ember-data@<3.11`.
```
Comment (runspired) [inline docs/rules/new-ember-data-packages.md]: just making a note to link any polyfill we create for older versions here in the future, nothing actionable at this time

File: lib/rules/new-ember-data-packages.js [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+
+//------------------------------------------------------------------------------
+// Rule Definition - Use "Ember Data Packages" from Ember RFC #395
+//------------------------------------------------------------------------------
+
+module.exports = {
+  meta: {
+    docs: {
+      description: 'Use "Ember Data Packages" from Ember RFC #395',
+      category: 'Best Practises',
```
Comment (runspired) [inline lib/rules/new-ember-data-packages.js]: s/Practises/Practices

File: lib/rules/new-ember-data-packages.js [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+    fixable: 'code',
+  },
+
+  create(context) {
+    //----------------------------------------------------------------------
+    // Variables
+    //----------------------------------------------------------------------
+
+    const message =
+      'Imports from new @ember-data packages should be prefered over imports from ember-data';
```
Comment (runspired) [inline lib/rules/new-ember-data-packages.js]: let's not use `new` here as in a few versions this will no longer be "new" :D

File: lib/rules/new-module-imports.js [hunk 4]
```diff
@@ -33,6 +33,24 @@ module.exports = {
   },
 
   create(context) {
+    function reportNestedProperties(properties, parent) {
+      properties.forEach(item => {
+        const match = GLOBALS[`Ember.${parent}.${item.key.name}`];
```
Comment (runspired) [inline lib/rules/new-module-imports.js]: is hard coding `Ember.` here correct when this is also used for data rules?

File: lib/rules/new-module-imports.js [hunk 5]
```diff
@@ -84,72 +107,15 @@ module.exports = {
           const key = fullName.replace(/^Ember\./, '');
```
Comment (runspired) [inline lib/rules/new-module-imports.js]: Do we need to also screen for `DS.` here?

File: tests/lib/rules/new-ember-data-packages.js [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+      code: `import Model from 'ember-data/model';
+
+        export default Model.extend({
+        });
+      `,
+      parserOptions,
+      errors: [
+        {
+          message:
+            'Imports from new @ember-data packages should be prefered over imports from ember-data',
```
Comment (runspired) [inline tests/lib/rules/new-ember-data-packages.js]: likely should drop use of "new" here as well

File: tests/lib/rules/new-ember-data-packages.js [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+
+        export default Model.extend({
+          name: DS.attr('string')
+        });
+      `,
+      parserOptions,
+      errors: [
+        {
+          message:
+            'Imports from new @ember-data packages should be prefered over imports from ember-data',
```
Comment (runspired) [inline tests/lib/rules/new-ember-data-packages.js]: likely should drop use of "new" here as well

File: tests/lib/rules/new-ember-data-packages.js [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
```
Comment (runspired) [inline tests/lib/rules/new-ember-data-packages.js]: likely should drop use of "new" here as well

File: tests/lib/rules/new-ember-data-packages.js [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+      code: `import DS from 'ember-data';
+
+        export default DS.Model.extend({
+        });
+      `,
+      parserOptions,
+      errors: [
+        {
+          message:
+            'Imports from new @ember-data packages should be prefered over imports from ember-data',
```
Comment (runspired) [inline tests/lib/rules/new-ember-data-packages.js]: likely should drop use of "new" here as well

File: tests/lib/rules/new-ember-data-packages.js [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+
+        export default Model.extend({
+          name: attr('string')
+        });
+      `,
+      parserOptions,
+      errors: [
+        {
+          message:
+            'Imports from new @ember-data packages should be prefered over imports from ember-data',
```
Comment (runspired) [inline tests/lib/rules/new-ember-data-packages.js]: likely should drop use of "new" here as well

Although we seem to be redeclaring this string a lot, maybe we should hoist it up to a module level const.

File: tests/lib/rules/new-ember-data-packages.js [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+          type: 'ImportDeclaration',
+        },
+        {
+          message:
+            'Imports from new @ember-data packages should be prefered over imports from ember-data',
```
Comment (runspired) [inline tests/lib/rules/new-ember-data-packages.js]: likely should drop use of "new" here as well

File: tests/lib/rules/new-ember-data-packages.js [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+
+        export default Model.extend({
+          name: attr('string')
+        });
+      `,
+      parserOptions,
+      errors: [
+        {
+          message:
+            'Imports from new @ember-data packages should be prefered over imports from ember-data',
```
Comment (runspired) [inline tests/lib/rules/new-ember-data-packages.js]: likely should drop use of "new" here as well

File: tests/lib/rules/new-ember-data-packages.js [hunk 13]
```diff
@@ -0,0 +1,185 @@
...
```
Comment (runspired) [inline tests/lib/rules/new-ember-data-packages.js]: likely should drop use of "new" here as well

File: tests/lib/rules/new-ember-data-packages.js [hunk 14]
```diff
@@ -0,0 +1,185 @@
...
+      code: `import { Model } from 'ember-data';
+
+        export default Model.extend({
+        });
+      `,
+      parserOptions,
+      errors: [
+        {
+          message:
+            'Imports from new @ember-data packages should be prefered over imports from ember-data',
```
Comment (runspired) [inline tests/lib/rules/new-ember-data-packages.js]: likely should drop use of "new" here as well

whew, sorry for like spamming you with a dozen of the same feedback

File: docs/rules/new-ember-data-packages.md (hunk 1 above)
Comment (dcyriller) [reply docs/rules/new-ember-data-packages.md]: 👍

File: lib/rules/new-ember-data-packages.js (hunk 3 above)
Comment (dcyriller) [reply lib/rules/new-ember-data-packages.js]: ahah yes indeed :)

File: lib/rules/new-module-imports.js (hunk 4 above)
Comment (dcyriller) [reply lib/rules/new-module-imports.js]: I've extracted some `new-module-imports` rule internals to `lib/utils/new-module.js`. This is why some lines change here. I did that to use the `buildMessage` function in the new rule `new-ember-data-packages` (btw, should we rename the rule as well or is it fine?).

So there are now two rules:
- `new-module-imports`: it applies "New Module Imports" from Ember RFC #176
- `new-ember-data-packages`: it applies "Ember Data Packages" from Ember RFC #395

Harcoding `Ember` is fine here. :)

The

File: tests/lib/rules/new-ember-data-packages.js (hunk 10 above)
Comment (dcyriller) [reply tests/lib/rules/new-ember-data-packages.js]: I've done that.

File: tests/lib/rules/new-ember-data-packages.js (hunk 14 above)
Comment (dcyriller) [reply tests/lib/rules/new-ember-data-packages.js]: ahah no worries :)

Comment (dcyriller) [reply tests/lib/rules/new-ember-data-packages.js]: I've fixed them

**Comment by dcyriller** [comment]
This is ready now.

File: docs/rules/new-ember-data-packages.md [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+import Model, { attr } from "@ember-data/model";
+
+import Model from "@ember-data/model";
+```
+
+## How to fix
+
+This rule implements a fix function (or to be more precise: it leverages the `no-old-shims` fix function). If, as a user, you can get rid of all the `import DS from "ember-data";` imports, you'll be able to use this rule fixer to complete the upgrade. 
+
+Note that a codemod is also available to complete the upgrade.
```
Comment (snewcomer) [inline docs/rules/new-ember-data-packages.md]: should we link the codemod here?

File: lib/rules/new-ember-data-packages.js [hunk 16]
```diff
@@ -0,0 +1,167 @@
...
+function globalsReducer(acc, mapping) {
+  const { global } = mapping;
+  if (global in acc) {
+    return acc;
+  }
+
+  let obj = {
+    global,
+    export: mapping.replacement.export,
+    localName: mapping.localName,
```
Comment (snewcomer) [inline lib/rules/new-ember-data-packages.js]: https://github.com/ember-data/ember-data-rfc395-data/blob/master/mappings.json#L154

Does `attr` need a localeName of `attr`?

File: lib/rules/new-ember-data-packages.js [hunk 17]
```diff
@@ -0,0 +1,167 @@
...
+    // Public
+    //----------------------------------------------------------------------
+
+    return {
+      /**
+       * turn `import Model from "ember-data/model"`
+       * into `import Model from "@ember-data/model"`
+       * or
+       * `import attr from "ember-data/attr"`
+       * into `import { attr } from "ember-data/attr"`
```
Comment (snewcomer) [inline lib/rules/new-ember-data-packages.js]: `import { attr } from "ember-data/model` ??

Comment (dcyriller) [reply lib/rules/new-ember-data-packages.js]: good catch 😉

File: docs/rules/new-ember-data-packages.md (hunk 15 above)
Comment (dcyriller) [reply docs/rules/new-ember-data-packages.md]: I add it

File: lib/rules/new-ember-data-packages.js (hunk 16 above)
Comment (dcyriller) [reply lib/rules/new-ember-data-packages.js]: Indeed! I've opened a PR to correct that: https://github.com/ember-data/ember-data-rfc395-data/pull/8

File: docs/rules/new-ember-data-packages.md (hunk 1 above)
Comment (runspired) [reply docs/rules/new-ember-data-packages.md]: link to issue: https://github.com/emberjs/data/issues/6266

File: lib/index.js [hunk 18]
```diff
@@ -13,6 +13,7 @@ module.exports = {
     'jquery-ember-run': require('./rules/jquery-ember-run'),
     'local-modules': require('./rules/local-modules'),
     'named-functions-in-promises': require('./rules/named-functions-in-promises'),
+    'new-ember-data-packages': require('./rules/new-ember-data-packages'),
```
Comment (rwjblue) [inline lib/index.js]: Sorry about this one, but I **really really really** dislike using "new" in the name of things (yes I know the `new-module-imports` has existed for a long time, but we can't really fix that since its already been released). Can we think up a different rule name here? Just brainstorming:

* require-rfc-395-ember-data-imports
* no-ember-data-global-imports
* no-ds-global

Have any better ideas?

File: lib/rules/new-ember-data-packages.js [hunk 19]
```diff
@@ -0,0 +1,168 @@
...
+          context.report(node, message);
+        }
+      },
+
+      /**
+       * warn against `const { Model } = DS`
+       */
+      VariableDeclarator(node) {
+        // Filter out non-DS variable declarations
+        if (!node.init || node.init.name !== 'DS') return;
```
Comment (rwjblue) [inline lib/rules/new-ember-data-packages.js]: Shouldn't this be tracked from imports as well as checking `DS` as is done now?

Can you add a test case for:

```js
import ED from 'ember-data';

const { Model } = ED;
```

File: lib/index.js (hunk 18 above)
Comment (dcyriller) [reply lib/index.js]: Indeed, I picked that name in reference to the `new-modules-imports` rule.

> I really really really dislike using "new" in the name of things 

😂 I can understand that. So called `new` things tend to be outdated pretty fast. Especially in the javascript / front-end world.

It could also be: `no-old-ember-data-imports`.

But `no-ember-data-global-imports` sounds better. I'll go for that!

Comment (rwjblue) [reply lib/index.js]: 👍 	sounds good!

File: lib/rules/new-ember-data-packages.js (hunk 19 above)
Comment (dcyriller) [reply lib/rules/new-ember-data-packages.js]: Thank you, I've added the test case and saw that your snippet could generate two warnings (instead of one). To do so, I've removed the line you pointed me to. Does it look to you? I'm not 100% sure I have understood what you meant. (I keep this modification in an isolated commit to rework it if needed.)

File: lib/index.js (hunk 18 above)
Comment (dcyriller) [reply lib/index.js]: It is renamed by now!

File: lib/rules/new-ember-data-packages.js (hunk 19 above)
Comment (rwjblue) [reply lib/rules/new-ember-data-packages.js]: New test looks great! Here is another to add:

```js
import LOL from 'who-knows-but-definitely-not-ember-data';

const { Model } = LOL;
```

I would not expect ^ to error/warn, but I _think_ the current code (as written) would.

File: docs/rules/no-ember-data-global-imports.md [hunk 20]
```diff
@@ -0,0 +1,41 @@
+# Use &#34;Ember Data Packages&#34; from Ember RFC #395 (no-ember-data-global-imports)
+
+The goal of this rule is to ease the migration to the new @ember-data packages.
+
+## Context
+
+`ember-data has been split in multiple packages. For instance, its store is now released in "@ember-data/store" package. These packages have been released starting from ember-data version 3.11.
```
Comment (bmish) [inline docs/rules/no-ember-data-global-imports.md]: nit: extra backtick

File: lib/index.js [hunk 21]
```diff
@@ -13,6 +13,7 @@ module.exports = {
     'jquery-ember-run': require('./rules/jquery-ember-run'),
     'local-modules': require('./rules/local-modules'),
     'named-functions-in-promises': require('./rules/named-functions-in-promises'),
+    'no-ember-data-global-imports': require('./rules/no-ember-data-global-imports'),
```
Comment (bmish) [inline lib/index.js]: Please sort alphabetically.

File: tests/lib/rules/no-ember-data-global-imports.js [hunk 22]
```diff
@@ -0,0 +1,196 @@
...
+const RuleTester = require('eslint').RuleTester;
+const parserOptions = { ecmaVersion: 6, sourceType: 'module' };
+
+const message = 'Imports from @ember-data packages should be prefered over imports from ember-data';
+
+//------------------------------------------------------------------------------
+// Tests
+//------------------------------------------------------------------------------
+
+const ruleTester = new RuleTester();
```
Comment (bmish) [inline tests/lib/rules/no-ember-data-global-imports.js]: Use `new RuleTester(parserOptions)` instead of passing `parserOptions` to each test case.

File: tests/lib/rules/no-ember-data-global-imports.js [hunk 23]
```diff
@@ -0,0 +1,196 @@
...
+    },
+    {
+      code: `import Model, { attr } from '@ember-data';
+
+        export default Model.extend({
+          name: attr('string')
+        });
+      `,
+      parserOptions,
+    },
```
Comment (bmish) [inline tests/lib/rules/no-ember-data-global-imports.js]: You can just have an array of strings for the `valid` test cases, no need to pass an object for each.

```
valid: [
  'import ...',
  'import ...'
]

File: lib/rules/no-ember-data-global-imports.js [hunk 24]
```diff
@@ -0,0 +1,165 @@
...
+
+//------------------------------------------------------------------------------
+// Rule Definition - Use "Ember Data Packages" from Ember RFC #395
+//------------------------------------------------------------------------------
+
+module.exports = {
+  meta: {
+    docs: {
+      description: 'Use "Ember Data Packages" from Ember RFC #395',
+      category: 'Best Practices',
```
Comment (bmish) [inline lib/rules/no-ember-data-global-imports.js]: It might be a good time for us to create a new `Ember Data` category. This rule along with the existing `no-empty-attrs` could go into it. If you make this change, remember to run `yarn update` afterward so the README gets updated.

File: tests/lib/rules/no-ember-data-global-imports.js (hunk 22 above)
Comment (dcyriller) [reply tests/lib/rules/no-ember-data-global-imports.js]: nice

File: lib/rules/no-ember-data-global-imports.js (hunk 24 above)
Comment (dcyriller) [reply lib/rules/no-ember-data-global-imports.js]: new category created

File: lib/rules/new-ember-data-packages.js (hunk 19 above)
Comment (dcyriller) [reply lib/rules/new-ember-data-packages.js]: I've addressed the issue in the last commit: https://github.com/ember-cli/eslint-plugin-ember/pull/450/commits/ca976143e2bd7c1c8d9408b6a423dbc8446922b0

File: lib/rules/no-ember-data-global-imports.js (hunk 24 above)
Comment (dcyriller) [reply lib/rules/no-ember-data-global-imports.js]: Also I’ve addressed all your comments!

File: tests/lib/rules/no-ember-data-global-imports.js [hunk 25]
```diff
@@ -0,0 +1,185 @@
...
+
+//------------------------------------------------------------------------------
+// Requirements
+//------------------------------------------------------------------------------
+
+const rule = require('../../../lib/rules/no-ember-data-global-imports');
+const RuleTester = require('eslint').RuleTester;
+const parserOptions = { ecmaVersion: 6, sourceType: 'module' };
+
+const message = 'Imports from @ember-data packages should be prefered over imports from ember-data';
```
Comment (bmish) [inline tests/lib/rules/no-ember-data-global-imports.js]: Instead of duplicating this error message here, we should import it from the rule file. Take a look at the `no-test-and-then` rule to see an example of how this importing can be done.

Comment (dcyriller) [reply tests/lib/rules/no-ember-data-global-imports.js]: This is done now

**Comment by rwjblue** [comment]
@bmish 	- Mind doing one more pass on this one before we land / release?

**Comment by runspired** [comment]
I dislike being pedantic especially when I feel the name has been changed a few times, but is `no-ember-data-global-imports` really the best name for a rule that:

- lints against older module import locations
- AND lints against an older "single-export-object" style import
- AND lints against global syntax

There was a suggestion to name the rule after the rfc, I think it's probably the way to go.

`use-ember-data-rfc-395-imports`

**Comment by dcyriller** [comment]
> There was a suggestion to name the rule after the rfc, I think it's probably the way to go. `use-ember-data-rfc-395-imports`

@rwjblue @bmish if you agree, I'll change that. Let me know! (I want to be sure everyone agrees before changing it.)

**Comment by rwjblue** [comment]
Yeah, that seems fine to me. My only real objection was the "new" in the original name, I'm sorry for the run around / bike shedding here.

**Comment by dcyriller** [comment]
No worries, I prefer to make it right!

File: lib/utils/new-module.js [hunk 26]
```diff
@@ -0,0 +1,146 @@
...
+const EMBER_NAMESPACES = ['inject.controller', 'inject.service'];
+
+function buildMessage(obj) {
+  // Both the Controller module and Service module each make available the
+  // "inject" namespace. In order to make it more readable, so it's more
+  // explicit at first glance from which module the "inject" namespace
+  // belongs to or is being imported from, let's check against this so we
+  // can properly populate the import specifier to report the ESLint error.
+  // Ex: import { inject as service } from '@ember/service';
+  const isNamespace = EMBER_NAMESPACES.indexOf(`${obj.parent}.${obj.key}`) !== -1;
```
Comment (bmish) [inline lib/utils/new-module.js]: Use `.includes()`

File: lib/rules/use-ember-data-rfc-395-imports.js [hunk 27]
```diff
@@ -0,0 +1,175 @@
...
+       * into `import { attr } from "@ember-data/attr"`
+       */
+      ImportDeclaration(node) {
+        if (node.source.value in OLD_DATA_IMPORTS) {
+          const fix = buildFix(node, OLD_DATA_IMPORTS);
+          context.report({ node, message, fix });
+          return;
+        }
+
+        if (node.source.value.match(/^ember-data/)) {
```
Comment (bmish) [inline lib/rules/use-ember-data-rfc-395-imports.js]: Use: `startsWith('ember-data/')`

File: lib/utils/new-module.js [hunk 28]
```diff
@@ -0,0 +1,146 @@
...
+ *   'ember-array/utils': {
+ *     A: ['@ember/array', 'A'],
+ *     isEmberArray: ['@ember/array', 'isArray'],
+ *     wrap: ['@ember/array', 'makeArray'],
+ *   },
+ * }
+ * @return {function(fixer: Object): function} a function that applies a fix to
+ * update the import
+ */
+function buildFix(node, modulesData) {
```
Comment (bmish) [inline lib/utils/new-module.js]: This function is too long and complicated and should have simple helper functions extracted from it.

File: lib/utils/new-module.js [hunk 29]
```diff
@@ -0,0 +1,146 @@
+'use strict';
+
+const EMBER_NAMESPACES = ['inject.controller', 'inject.service'];
+
+function buildMessage(obj) {
```
Comment (bmish) [inline lib/utils/new-module.js]: This function is too long and complicated and should have simple helper functions extracted from it.

File: lib/rules/use-ember-data-rfc-395-imports.js [hunk 30]
```diff
@@ -0,0 +1,176 @@
...
+          }
+        });
+      },
+
+      /**
+       * warn against DS usage in `name: DS.attr('string')`
+       * taken from
+       * https://github.com/ember-cli/eslint-plugin-ember/blob/v6.7.1/lib/rules/new-module-imports.js#L63
+       */
+      'MemberExpression > Identifier[name=DS]'(node) {
```
Comment (bmish) [inline lib/rules/use-ember-data-rfc-395-imports.js]: This function is too long and complicated and should have simple helper functions extracted from it.

File: lib/utils/new-module.js (hunk 26 above)
Comment (dcyriller) [reply lib/utils/new-module.js]: This is [existing code on master](https://github.com/ember-cli/eslint-plugin-ember/blob/v6.7.1/lib/rules/new-module-imports.js#L124). It is part of a function I've extracted to an util. The goal is to reuse it between two rules. With that explanation, do you still think I should implement your suggestion?

File: lib/utils/new-module.js (hunk 28 above)
Comment (dcyriller) [reply lib/utils/new-module.js]: This function is [existing on master](https://github.com/ember-cli/eslint-plugin-ember/blob/v6.7.1/lib/rules/no-old-shims.js#L273), as a part of a longer function. I've extracted to an util to be reused between two rules. With that explanation, do you think I should implement your suggestion?

File: lib/utils/new-module.js (hunk 29 above)
Comment (dcyriller) [reply lib/utils/new-module.js]: This is existing (and longer) code that I've extracted to an util to be reused between two rules. With that explanation, do you think I should implement your suggestion?

**Comment by bmish** [comment]
@dcyriller don't worry, I'm not going to require you to refactor existing code if you're just moving it around. I commented about the excessive length and complexity of some of the functions because it makes the review difficult, and it could be a good time to clean this up. We can merge the PR soon if you're done with your changes.

**Comment by dcyriller** [comment]
@bmish should be good now.

**Comment by dcyriller** [comment]
I've made fun splitting utils functions into smaller ones. I'm not 100% convinced it makes the code better though.

File: .travis.yml [hunk 31]
```diff
@@ -7,8 +7,8 @@ node_js:
 cache: yarn
 
 script:
-  - yarn run lint
-  - yarn run test:coverage --runInBand
+  - yarn lint
+  - yarn test:coverage --runInBand
```
Comment (bmish) [inline .travis.yml]: @dcyriller why is this changing now? This should be in a separate PR if we want to change it.

**Comment by bmish** [comment]
@rwjblue let's merge this so we can release a new version today?

**Comment by bendemboski** [comment]
@dcyriller @bmish @rwjblue this caused a [crashing regression](https://github.com/ember-cli/eslint-plugin-ember/issues/459). Let me know if there's any way I can help get this resolved as it's blocking me updating to the latest `eslint-plugin-ember`.
